### PR TITLE
Uninstall unused kernels

### PR DESCRIPTION
**Pull request by framps**
This PR is the outcome of [this issue ](https://github.com/RPi-Distro/raspi-config/issues/244) and based on [this script](https://github.com/framps/raspberryTools/blob/master/raspiHandleKernels.sh) I wrote to uninstall and reinstall unused kernels to reduce the apt upgrade time because otherwise all possible kernels are rebuilt even there is only one kernel used by a system. This is no issue on an RPi5 but a serious issue on an RPi1.

### CODE REVIEW

**Comment by XECDesign** [comment]
A couple of problems which break the update and the autoremove mechanisms...

1) If you run apt full-upgrade and get a new kernel, then run this, it will uninstall your new kernel and only keep the now-outdated kernel.
2) It treats the meta-packages and actual kernel packages the same way, which leads it to removing all the meta-packages each time, meaning you no longer get kernel updates.
3) On re-install the kernel will be apt-marked as 'manual', meaning apt won't know that it needs to clean them up after updates

I'll push an alternative approach commit to this PR tomorrow.

**Comment by XECDesign** [comment]
A few more (getting really nit-picky):

1) It's not showing up in the menu on a pi5
2) It should probably be in the advanced options menu instead of performance
3) If a kernel is installed between runs, do_kernels will remove it and overwrite the unused kernels file, losing the previously removed kernels
4) Use of `RET` seems strange. There's nothing that returns it at the end and in the nonint case, it's set to $1?
5) The unused kernels file is in a non-FHS location
6) The unused kernels file is left behind when the raspi-config package is removed.
7) Inconsistent indentation
8) Shellcheck warnings

In general, the re-install option seems a bit error-prone and handling it properly is starting to look very messy.

Haven't had a chance to tidy things up fully yet. Will re-visit when I get a chance.

**Comment by framps** [comment]
> getting really nit-picky

No worries. Its a feature which has to be worked on carefully. Thank you very much for your comments and review. 

Regarding your previous comments about meta packages I have to admit I know the package handling from a users perspective but not from a development perspective and need some guidance to get it right.

> 1. It's not showing up in the menu on a pi5

At the time I created the PR I didn't have a RPi5. In the meantime I got one donated by Jeff. Will fix this. :white_check_mark:

> 2. It should probably be in the advanced options menu instead of performance

It's not a runtime performance improvement but an apt update performance improvement. I agree the advanced options menu is a better place. Will change this. :white_check_mark:

> 3. If a kernel is installed between runs, do_kernels will remove it and overwrite the unused kernels file, losing the previously removed kernels

Good catch. Will check and fix.

> 4. Use of `RET` seems strange. There's nothing that returns it at the end and in the nonint case, it's set to $1? :white_check_mark:

Will check

> 5. The unused kernels file is in a non-FHS location

This can be easily be changed. But I have no clue what will be the correct FHS location for this. 

> 6. The unused kernels file is left behind when the raspi-config package is removed.

Good catch. But as far as I can see that's something which has to be managed in the raspi-config package, doesn't it? As I wrote already - I don't have package handling knowledge from a development side and need help on this. I frankly didn't expect I have to.

> 7. Inconsistent indentation

I'm used to indent consistently. Will check and fix. :white_check_mark:

> 8. Shellcheck warnings

I recently learned this nice tool exists. Will remove the warnings. :white_check_mark:

**Comment by XECDesign** [comment]
> Regarding your previous comments about meta packages I have to admit I know the package handling from a users perspective but not from a development perspective and need some guidance to get it right.

Apt has the concept of 'manual' and 'auto' packages:

```
$ apt-mark showmanual linux-image\*
linux-image-rpi-2712
$ apt-mark showauto linux-image\*
linux-image-6.6.62+rpt-rpi-2712
```

'auto' means it was pulled in as a dependency of something else. 'manual' means you've explicitly installed it yourself.

If nothing on the system depends on an 'auto' package, apt will tell you it can be removed. In the case of kernels, `apt full-upgrade` should automatically remove everything but the last two versions.

As is, this tool will always remove the meta-packages (linux-image-rpi-2712, for example). Then when we ship a newer kernel, apt won't know that it needs to install it.

Also, let's say the tool has removed 'linux-image-6.6.62+rpt-rpi-2712' and then re-installed it, it will be marked 'manual', so it will never be a candidate for automatic removal even after it's no longer being used and several newer versions have shipped.

To demonstrate,  I'll install some kernels:

```
$ sudo apt install linux-image-6.6.51+rpt-rpi-2712 linux-image-6.6.47+rpt-rpi-2712 linux-image-6.6.31+rpt-rpi-2712
...
$ linux-version list
6.6.31+rpt-rpi-2712
6.6.47+rpt-rpi-2712
6.6.51+rpt-rpi-2712
6.6.62+rpt-rpi-2712
$ sudo apt full-upgrade
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
```

In this case, full-upgrade does nothing, which is not the desired behaviour. 

If we mark them as 'auto', as they would normally be...

```
$ sudo apt-mark auto linux-image-6.6.51+rpt-rpi-2712 linux-image-6.6.47+rpt-rpi-2712 linux-image-6.6.31+rpt-rpi-2712
linux-image-6.6.51+rpt-rpi-2712 set to automatically installed.
linux-image-6.6.47+rpt-rpi-2712 set to automatically installed.
linux-image-6.6.31+rpt-rpi-2712 set to automatically installed.
shift@serge-testpi ~> sudo apt full-upgrade
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
The following packages will be REMOVED:
  linux-image-6.6.31+rpt-rpi-2712 linux-image-6.6.47+rpt-rpi-2712
0 upgraded, 0 newly installed, 2 to remove and 0 not upgraded.
After this operation, 66.2 MB disk space will be freed.
Do you want to continue? [Y/n] n
```

It will detect the old kernels and automatically remove them.

If we remove the meta-package and the latest kernel version, leaving just the old versions...

```
$ sudo apt purge linux-image-6.6.62+rpt-rpi-2712 linux-image-rpi-2712
$ sudo apt full-upgrade --auto-remove --purge
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
The following packages will be REMOVED:
  linux-image-6.6.31+rpt-rpi-2712*
0 upgraded, 0 newly installed, 1 to remove and 0 not upgraded.
After this operation, 32.9 MB disk space will be freed.
Do you want to continue? [Y/n] n
Abort.
```

Then apt is no  longer aware of the fact there's a newer kernel available. Reinstalling the meta-package restores the desired behaviour:

```
$ sudo apt install linux-image-rpi-2712
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following package was automatically installed and is no longer required:
  linux-image-6.6.31+rpt-rpi-2712
Use 'sudo apt autoremove' to remove it.
The following additional packages will be installed:
  linux-image-6.6.62+rpt-rpi-2712
Suggested packages:
  linux-doc-6.6 debian-kernel-handbook
The following NEW packages will be installed:
  linux-image-6.6.62+rpt-rpi-2712 linux-image-rpi-2712
0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 29.7 MB of archives.
After this operation, 33.5 MB of additional disk space will be used.
Do you want to continue? [Y/n] n
Abort.
```

To sum up, for apt to handle kernel updates correctly, the meta-package needs to me installed and marked 'manual', the packages containing the actual kernel need to be marked 'auto'.

Hope that makes sense and wasn't too waffly.

**Comment by XECDesign** [comment]
> This can be easily be changed. But I have no clue what will be the correct FHS location for this.

I'd recommend abandoning that particular part of it, but if you'd like to keep it, I think [/var/lib/raspi-config/](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html) makes the most sense.

I can add the packaging bit to remove the file when the package is purged. Not a likely scenario, but will keep QC tools which look for packages that leave cruft behind happier.

**Comment by XECDesign** [comment]
For reference, I have this:

```shell
do_kernels() {

  if [ "$INTERACTIVE" != True ]; then
    echo "nonint mode unsupported by do_kernels" 2>&1
    return 1
  fi
  DEFAULT=--defaultno

  KERNELS_AVAILABLE="$(linux-version list | linux-version sort --reverse)"
  KERNEL_RUNNING="$(uname -r)"
  KERNEL_FAM="$(echo "$KERNEL_RUNNING" | rev | cut -f1 -d- | rev)"

  case "$KERNEL_FAM" in
  2712|v8|v7l|v7|v6)
    ;;
  *)
    whiptail --msgbox "Unrecognised kernel family ($KERNEL_FAM)" 20 60 1
    return 1
    ;;
  esac

  KERNEL_LATEST="$(echo "$KERNELS_AVAILABLE" | grep -m1 "\-$KERNEL_FAM\$")"
  KERNELS_USED="$(printf '%s\n%s\n' "$KERNEL_RUNNING" "$KERNEL_LATEST" | uniq)"
  KERNELS_UNUSED="$KERNELS_AVAILABLE"
  for used in $KERNELS_USED; do
    KERNELS_UNUSED="$(echo "$KERNELS_UNUSED" | sed "/^$used\$/d")"
  done
  KERNELS_UNUSED_PACKAGES="$(echo "$KERNELS_UNUSED" | sed 's/^./linux-image-&/')"

  if [ -n "$KERNELS_UNUSED_PACKAGES" ]; then
    for package in $KERNELS_UNUSED_PACKAGES; do
      if ! is_installed "$package"; then
        whiptail --msgbox "Installed kernel does not match a package ($package)" 20 60 1
        return 1
      fi
    done
    whiptail --yesno \
      "Are you sure you would like $(echo "$KERNELS_UNUSED_PACKAGES" | wc -l) unused kernel(s) to be uninstalled?\n\nThis will break your ability to boot this install on other pi models.\n\nIf you are unsure, say No." \
      $DEFAULT 20 60 2
      RET=$?
      if [ $RET -eq 0 ]; then
        if ! echo "$KERNELS_UNUSED_PACKAGES" | xargs -r apt-get -y remove; then
          whiptail --msgbox "Kernel uninstallation failed." 20 60 1
          return 1
        fi
      fi
  else
    whiptail --msgbox "No unused kernels found." 20 60 1
  fi
  return 0
}
```

This should error out when run in a chroot environment or if some unexpected kernel is installed.

**Comment by framps** [comment]
> I'd recommend abandoning that particular part of it, but if you'd like to keep it

If this file is dropped there is no easy way to get a stripped down system to boot on a different RPi HW which requires a different kernel than the one kept. Not sure how likely it's somebody needs this functionality and it's worth to keep the old kernels. On the other side there is no need on the on the packaging side to remove this file.

**Comment by framps** [comment]
@XECDesign I just noticed there were a lot of commits added which I have to either rebase or merge in this PR. I started a rebase but that's a lot of work. A merge just needs to manually merge only two files:  `raspi-config` and `debian/control`. Given the amount of changes and I'm not familiar with all the code changes I prefer to merge the new commits.

Do you want me to rebase or can I just merge? If I should rebase I will start over and create a new PR.

**Comment by XECDesign** [comment]
Heads up you don't need to create a new PR, you can always force-push to this one. I was going to rebase this, but it seems to be in a "better start from scratch" state. Or at least I am not good enough at git to detangle this.

You can save your changes somewhere else, run `git reset origin/bookworm` to get back to the upstream state, reapply and re-commit your changes, then run `git push -f` to update this PR.

**Comment by framps** [comment]
It took me a while to get git to do what you proposed to do. But finally I manged to get it done :smile: Thank you very much for the guidance. I didn't realize a force push will be executed on my repo only which is OK but not on yours.

I frankly just c&p you proposed code for do_kernels and it works like a charm :+1: 

When I tested the code on my rpi5 I got a message `Installed kernel does not match a package`. Root cause is 

```
pi@raspberrypi-bookworm64-lite-beta:~ $ dpkg -l linux-image-6.6.62+rpt-rpi-v8
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                          Version         Architecture Description
+++-=============================-===============-============-===========================================
iF  linux-image-6.6.62+rpt-rpi-v8 1:6.6.62-1+rpt1 arm64        Linux 6.6 for Raspberry Pi v8, Raspberry Pi
```
and
```
is_installed() {
  if [ "$(dpkg -l "$1" 2> /dev/null | tail -n 1 | cut -d ' ' -f 1)" != "ii" ]; then
    return 1
  else
    return 0
  fi
}
```
checks for `ii`but rejects `iF`. My other Raspberries - RPi1B and RPi4 - work fine with do_kernels. 

I also updated my raspiHandleKernels script in my raspberryTools repository with your comments and findings in mind. I still keep there the ability to reinstall the uninstalled kernels. But I agree it's unlikely somebody want's to reinstall an uninstalled kernel. And if it's required there are means to install the required kernel again. But it's not convenient for non experienced users.

**Comment by framps** [comment]
@XECDesign 

I just noticed there were updates on upstream and I have to resolve conflicts in my PR. 

I'm now wondering whether it's worth to spend any time on this PR to resolve conflicts given it's pending now for a long time.

**Comment by framps** [comment]
I tried to resolve conflicts which went wrong - unfortunately - and git decided to close this PR :cry: 

@XECDesign If you still are interested in this PR please open it again :wink:
